dahuffman/huffmancodec.py

```diff
@@ -64,58 +65,107 @@ def ensure_dir(path: Union[str, Path]) -> Path:
     return path
 
 
+class CodeTable:
+    """
+    Code table: mapping a symbol to codewords (and vice versa).
+
+    The symbols are the things you want to encode, usually characters in a string
+    or byte sequence, but it can be anything hashable.
+    The codewords are the corresponding bit sequences, represented as a tuple (bits, value)
+    where `bits` is the number of bits and `value` the integer interpretation of these bits.
+    """
+
+    # TODO: use something like namedtuple or class with slots for codewords instead of tuples?
+
+    def __init__(self, symbol_code_map: dict):
+        self._symbol_map = {}
+        self._code_map = {}
+        for symbol, (bits, value) in symbol_code_map.items():
+            if not (
+                isinstance(bits, int)
+                and bits >= 1
+                and isinstance(value, int)
+                and value >= 0
+                and math.log2(max(value, 1)) < bits
+            ):
+                raise ValueError(
+                    "Invalid code: {b} bits, value {v}".format(b=bits, v=value)
+                )
+            self._symbol_map[symbol] = (bits, value)
+            self._code_map[(bits, value)] = symbol
+        # TODO check if code table is actually a prefix code
+
+    def __len__(self) -> int:
+        return len(self._symbol_map)
+
+    def get_code(self, symbol: Any) -> Tuple[int, int]:
```

Comment:
Don't you think that this return type annotation should be put into a variable and reused?